GCC 9: new warnings for -Werror=address-of-packed-member option

Using current GCC trunk, one can see multiple errors:

```
[   79s] /home/abuild/rpmbuild/BUILD/ucx-1.4.0/src/ucs/datastruct/queue.h:269:22: error: taking address of packed member of 'struct <anonymous>' may result in an unaligned pointer value [-Werror=address-of-packed-member]
[   79s]   269 |     ucs_container_of(*iter, typeof(*elem), member)
[   79s] /home/abuild/rpmbuild/BUILD/ucx-1.4.0/src/ucs/sys/compiler_def.h:119:32: note: in definition of macro 'ucs_container_of'
[   79s]   119 |     ( (_type*)( (char*)(void*)(_ptr) - ucs_offsetof(_type, _member) )  )
[   79s]       |                                ^~~~
[   79s] ib/ud/base/ud_ep.c:900:16: note: in expansion of macro 'ucs_queue_iter_elem'
[   79s]   900 |     sent_skb = ucs_queue_iter_elem(sent_skb, ep->resend.pos, queue);
[   79s]       |                ^~~~~~~~~~~~~~~~~~~
[   79s] ib/ud/base/ud_ep.c:912:56: error: taking address of packed member of 'struct <anonymous>' may result in an unaligned pointer value [-Werror=address-of-packed-member]
[   79s]   912 |         ep->resend.pos = ucs_queue_iter_next(ep->resend.pos);
[   79s]       |                                              ~~~~~~~~~~^~~~
[   79s] ib/ud/base/ud_ep.c:925:52: error: taking address of packed member of 'struct <anonymous>' may result in an unaligned pointer value [-Werror=address-of-packed-member]
[   79s]   925 |     ep->resend.pos = ucs_queue_iter_next(ep->resend.pos);
[   79s]       |                                          ~~~~~~~~~~^~~~
[   79s] libtool: compile:  gcc -DHAVE_CONFIG_H -I. -I../.. -DCPU_FLAGS= -I/home/abuild/rpmbuild/BUILD/ucx-1.4.0/src -I/home/abuild/rpmbuild/BUILD/ucx-1.4.0/src -O3 -g -Wall -Werror -O2 -Wall -D_FORTIFY_SOURCE=2 -fstack-protector-strong -funwind-tables -fasynchronous-unwind-tables -fstack-clash-protection -g -c sm/base/sm_iface.c  -fPIC -DPIC -o sm/base/.libs/libuct_la-sm_iface.o
[   79s] /bin/sh ../../libtool  --tag=CC   --mode=compile gcc -DHAVE_CONFIG_H -I. -I../..  -DCPU_FLAGS="" -I/home/abuild/rpmbuild/BUILD/ucx-1.4.0/src -I/home/abuild/rpmbuild/BUILD/ucx-1.4.0/src       -O3 -g -Wall -Werror  -O2 -Wall -D_FORTIFY_SOURCE=2 -fstack-protector-strong -funwind-tables -fasynchronous-unwind-tables -fstack-clash-protection -g -c -o sm/base/libuct_la-sm_ep.lo `test -f 'sm/base/sm_ep.c' || echo './'`sm/base/sm_ep.c
[   79s] /bin/sh ../../libtool  --tag=CC   --mode=compile gcc -DHAVE_CONFIG_H -I. -I../..  -DCPU_FLAGS="" -I/home/abuild/rpmbuild/BUILD/ucx-1.4.0/src -I/home/abuild/rpmbuild/BUILD/ucx-1.4.0/src       -O3 -g -Wall -Werror  -O2 -Wall -D_FORTIFY_SOURCE=2 -fstack-protector-strong -funwind-tables -fasynchronous-unwind-tables -fstack-clash-protection -g -c -o sm/mm/libuct_la-mm_iface.lo `test -f 'sm/mm/mm_iface.c' || echo './'`sm/mm/mm_iface.c
[   79s] libtool: compile:  gcc -DHAVE_CONFIG_H -I. -I../.. -DCPU_FLAGS= -I/home/abuild/rpmbuild/BUILD/ucx-1.4.0/src -I/home/abuild/rpmbuild/BUILD/ucx-1.4.0/src -O3 -g -Wall -Werror -O2 -Wall -D_FORTIFY_SOURCE=2 -fstack-protector-strong -funwind-tables -fasynchronous-unwind-tables -fstack-clash-protection -g -c sm/base/sm_ep.c  -fPIC -DPIC -o sm/base/.libs/libuct_la-sm_ep.o
[   79s] libtool: compile:  gcc -DHAVE_CONFIG_H -I. -I../.. -DCPU_FLAGS= -I/home/abuild/rpmbuild/BUILD/ucx-1.4.0/src -I/home/abuild/rpmbuild/BUILD/ucx-1.4.0/src -O3 -g -Wall -Werror -O2 -Wall -D_FORTIFY_SOURCE=2 -fstack-protector-strong -funwind-tables -fasynchronous-unwind-tables -fstack-clash-protection -g -c sm/mm/mm_iface.c  -fPIC -DPIC -o sm/mm/.libs/libuct_la-mm_iface.o
[   79s] sm/mm/mm_iface.c: In function 'uct_mm_iface_create_signal_fd':
[   79s] sm/mm/mm_iface.c:433:42: error: taking address of packed member of 'struct uct_mm_fifo_ctl' may result in an unaligned pointer value [-Werror=address-of-packed-member]
[   79s]   433 |                       (struct sockaddr *)&iface->recv_fifo_ctl->signal_sockaddr,
[   79s]       |                                          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

--- a/src/ucp/stream/stream.h
+++ b/src/ucp/stream/stream.h
@@ -22,7 +22,7 @@ typedef struct {
         ucp_stream_am_hdr_t  hdr;
         ucp_recv_desc_t     *rdesc;
     };
-} UCS_S_PACKED ucp_stream_am_data_t;
+} ucp_stream_am_data_t;
 
 
 void ucp_stream_ep_init(ucp_ep_h ep);

--- a/src/uct/ib/ud/base/ud_def.h
+++ b/src/uct/ib/ud/base/ud_def.h
@@ -26,6 +26,7 @@
 #define UCT_UD_CA_MD_FACTOR     2   /* window = window/factor */
 #define UCT_UD_CA_DUP_ACK_CNT   2   /* TODO: not implemented yet */
 #define UCT_UD_RESENDS_PER_ACK  4   /* request per every N resends */
+#define UCT_UD_SKB_ALIGN        UCS_SYS_CACHE_LINE_SIZE
 
 /* note that the ud tx window is [acked_psn+1, max_psn)
  * and max_psn = acked_psn + cwnd
@@ -146,7 +147,7 @@ typedef struct uct_ud_send_skb {
     uint8_t                 flags;
     int8_t                  status;     /* used in case of failure */
     uct_ud_neth_t           neth[0];
-} UCS_S_PACKED UCS_V_ALIGNED(UCS_SYS_CACHE_LINE_SIZE) uct_ud_send_skb_t;
+} UCS_S_PACKED UCS_V_ALIGNED(UCT_UD_SKB_ALIGN) uct_ud_send_skb_t;
 
 
 typedef struct uct_ud_comp_desc {

--- a/src/uct/ib/ud/base/ud_ep.c
+++ b/src/uct/ib/ud/base/ud_ep.c
@@ -914,12 +914,20 @@ static uct_ud_send_skb_t *uct_ud_ep_resend(uct_ud_ep_t *ep)
 {
     uct_ud_iface_t *iface = ucs_derived_of(ep->super.super.iface, uct_ud_iface_t);
     uct_ud_send_skb_t *skb, *sent_skb;
+    ucs_queue_iter_t resend_pos;
     uct_ud_zcopy_desc_t *zdesc;
     size_t iov_it;
 
     /* check window */
-    sent_skb = ucs_queue_iter_elem(sent_skb, ep->resend.pos, queue);
-    if ((sent_skb == NULL) || UCT_UD_PSN_COMPARE(sent_skb->neth->psn, >=, ep->tx.max_psn)) {
+    resend_pos = (void*)ep->resend.pos;
+    sent_skb   = ucs_queue_iter_elem(sent_skb, resend_pos, queue);
+    if (sent_skb == NULL) {
+        uct_ud_ep_ctl_op_del(ep, UCT_UD_EP_OP_RESEND);
+        return NULL;
+    }
+
+    ucs_assert(((uintptr_t)sent_skb % UCT_UD_SKB_ALIGN) == 0);
+    if (UCT_UD_PSN_COMPARE(sent_skb->neth->psn, >=, ep->tx.max_psn)) {
         ucs_debug("ep(%p): out of window(psn=%d/max_psn=%d) - can not resend more",
                   ep, sent_skb ? sent_skb->neth->psn : -1, ep->tx.max_psn);
         uct_ud_ep_ctl_op_del(ep, UCT_UD_EP_OP_RESEND);
@@ -930,7 +938,7 @@ static uct_ud_send_skb_t *uct_ud_ep_resend(uct_ud_ep_t *ep)
     if ((uct_ud_neth_get_dest_id(sent_skb->neth) == UCT_UD_EP_NULL_ID) &&
         !(sent_skb->neth->packet_type & UCT_UD_PACKET_FLAG_CTL))
     {
-        ep->resend.pos = ucs_queue_iter_next(ep->resend.pos);
+        ep->resend.pos = ucs_queue_iter_next(resend_pos);
         return NULL;
     }
 
@@ -943,7 +951,7 @@ static uct_ud_send_skb_t *uct_ud_ep_resend(uct_ud_ep_t *ep)
     skb = uct_ud_iface_resend_skb_get(iface);
     ucs_assert_always(skb != NULL);
 
-    ep->resend.pos = ucs_queue_iter_next(ep->resend.pos);
+    ep->resend.pos = ucs_queue_iter_next(resend_pos);
     ep->resend.psn = sent_skb->neth->psn;
     memcpy(skb->neth, sent_skb->neth, sent_skb->len);
     skb->neth->ack_psn = ep->rx.acked_psn;

--- a/src/uct/ib/ud/base/ud_ep.h
+++ b/src/uct/ib/ud/base/ud_ep.h
@@ -240,9 +240,9 @@ struct uct_ud_ep {
         uct_ud_psn_t  cwnd;
     } ca;
     struct UCS_S_PACKED {
+         ucs_queue_iter_t       pos;       /* points to the part of tx window that needs to be resent */
          uct_ud_psn_t           psn;       /* last psn that was retransmitted */
          uct_ud_psn_t           max_psn;   /* max psn that should be retransmitted */
-         ucs_queue_iter_t       pos;       /* points to the part of tx window that needs to be resent */
     } resend;
     ucs_list_link_t  cep_list;
     uint32_t         conn_id;      /* connection id. assigned in connect_to_iface() */

--- a/src/uct/ib/ud/base/ud_iface.c
+++ b/src/uct/ib/ud/base/ud_iface.c
@@ -467,7 +467,7 @@ UCS_CLASS_INIT_FUNC(uct_ud_iface_t, uct_ud_iface_ops_t *ops, uct_md_h md,
     status = uct_iface_mpool_init(&self->super.super, &self->tx.mp,
                                   sizeof(uct_ud_send_skb_t) + data_size,
                                   sizeof(uct_ud_send_skb_t),
-                                  UCS_SYS_CACHE_LINE_SIZE,
+                                  UCT_UD_SKB_ALIGN,
                                   &config->super.tx.mp, self->config.tx_qp_len,
                                   uct_ud_iface_send_skb_init, "ud_tx_skb");
     if (status != UCS_OK) {

--- a/src/uct/sm/mm/base/mm_ep.c
+++ b/src/uct/sm/mm/base/mm_ep.c
@@ -87,6 +87,9 @@ static UCS_CLASS_INIT_FUNC(uct_mm_ep_t, uct_iface_t *tl_iface,
     self->signal.addrlen  = self->fifo_ctl->signal_addrlen;
     self->signal.sockaddr = self->fifo_ctl->signal_sockaddr;
 
+    /* Make sure the fifo ctrl is aligned */
+    ucs_assert_always(((uintptr_t)self->fifo_ctl % UCS_SYS_CACHE_LINE_SIZE) == 0);
+
     /* set the ep->fifo ptr to point to the beginning of the fifo elements at
      * the remote peer */
     uct_mm_set_fifo_elems_ptr(self->mapped_desc.address, &self->fifo);

--- a/src/uct/sm/mm/base/mm_iface.h
+++ b/src/uct/sm/mm/base/mm_iface.h
@@ -49,7 +49,7 @@ struct uct_mm_fifo_ctl {
 
     /* 2nd cacheline */
     volatile uint64_t  tail;       /* how much was read */
-} UCS_S_PACKED;
+} UCS_S_PACKED UCS_V_ALIGNED(UCS_SYS_CACHE_LINE_SIZE);
 
 
 struct uct_mm_iface {

